runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java

```diff
@@ -0,0 +1,294 @@
+package org.corfudb.runtime.collections;
+
+import com.google.common.collect.Streams;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.util.serializer.ISerializer;
+import org.rocksdb.RocksDB;
+import org.rocksdb.Options;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+/**
+ * A concrete implementation of {@link StreamingMap} that is capable of storing data
+ * off-heap. The location for the off-heap data is provided by {@link File} dataPath,
+ * while the resource policy (memory and storage limits) are defined in {@link Options}.
+ *
+ * @param <K> key type
+ * @param <V> value type
+ */
+@Slf4j
+public class PersistedStreamingMap<K, V> implements StreamingMap<K, V> {
+
+    private final AtomicInteger dataSetSize = new AtomicInteger();
+    private final CorfuRuntime corfuRuntime;
+    private final ISerializer serializer;
+    private final RocksDB rocksDb;
+
+    public PersistedStreamingMap(@NonNull File dataPath,
+                                 @NonNull Options options,
+                                 @NonNull ISerializer serializer,
+                                 @NonNull CorfuRuntime corfuRuntime) {
+        try {
+            FileUtils.deleteDirectory(dataPath);
+            this.rocksDb = RocksDB.open(options, dataPath.getAbsolutePath());
+        } catch (RocksDBException | IOException e) {
+            throw new UnrecoverableCorfuError(e);
+        }
+        this.serializer = serializer;
+        this.corfuRuntime = corfuRuntime;
+    }
+
+    /**
+     * A Java compatible {@link RocksIterator} implementation
+     */
+    public class RocksDbIterator implements Iterator<Entry<K, V>> {
+        private RocksIterator iterator;
```

Comment:
should these fileds be atomic or volatile?

Reply:
`RocksIterator` does not support multi-threaded access.

Reply:
why not? if rocks provides that, why not to support that

Reply:
`RocksIterator` is RocksDB implementaion. RocksDB does not support it. Not us.